src/Matrix.jl

```diff
@@ -1508,9 +1508,43 @@ julia> B = transpose(A)
 
 ```
 """
-transpose(x::MatrixElem{T}) where T <: NCRingElement
+function transpose(x::MatrixElem)
+  z = similar(base_ring(x), ncols(x), nrows(x))
+  return transpose!(z, x)
+end
+
+@doc raw"""
+    transpose!(x::MatrixElem)
+    transpose!(z::T, x::T) where T <: MatrixElem
+
+Return the transpose of `x`, possibly modifying the object `z` in the process.
+Aliasing is permitted.
+The unary version may modify `x` in-place (but this is not always possible).
+"""
+function transpose!(x::MatrixElem)
+  if is_square(x)
+    # square matrix, so we can transpose it in-place
+    n = nrows(x)
+    for i in 1:n
+      for j in i+1:n
+        x[i, j], x[j, i] = x[j, i], x[i, j]
+      end
+    end
+    return x
+  end
+  # we have no generic way to change the dimensions of x, so instead we
+  # delegate to the two argument version of transpose!
```

Comment:
Tests fails because in Nemo some `transpose!` methods with just one argument only work on square matrices.

Perhaps that the right decision, and I should match it here, and just throw an error in that case, too? Perhaps a bit more helpful one, but an error nevertheless.

Or perhaps the Nemo methods should be made more general.

Any thoughts on this, anyone?

Reply:
I never missed it for non-square matrices, but maybe @fieker has opinion/wishes?

Reply:
if it was missing - how can there be a test for it? I never used inplace transpose on non-square matrices. i do understand that this is possible (in flint) and an interesting research problem....